File: plugins/nonUniqueName.cpp
#include "medic/pluginapi.h"
#include "maya/MFn.h"
#include "maya/MFnDependencyNode.h"


using namespace MEDIC;


class NonUniqueName : public MdTester
{
    public:
        NonUniqueName();
        ~NonUniqueName();
        std::string Name();
        std::string Description();
        bool Match(MdNode *node);
        bool isFixable() const;
        MdReport *test(MdNode *node) const;
        MdParamContainer *GetParameters() const;
        bool Fix(MdReport *report) const;
};


NonUniqueName::NonUniqueName() {}

NonUniqueName::~NonUniqueName() {}

std::string NonUniqueName::Name()
{
    return "NonUniqueName";
}

std::string NonUniqueName::Description()
{
    return "Non unique name(s) exists";
}

bool NonUniqueName::Match(MdNode *node)
{
    return node->object().hasFn(MFn::kDagNode);
}

bool NonUniqueName::isFixable() const
{
    return true;
}

MdReport *NonUniqueName::test(MdNode *node) const
{
    MFnDependencyNode dg(node->object());
    if (dg.hasUniqueName())
    {
        return 0;
    }

    return new MdReport(node);
}

MdParamContainer *NonUniqueName::GetParameters() const
{
    MdParamContainer *params = new MdParamContainer();
    MdParameter *pattern = new MdParameter("pattern", "Rename Pattern", MdString, "");
    params->append(pattern);

    return params;
}

bool NonUniqueName::Fix(MdReport *report) const
{
    return true;
}


MEDIC_PLUGIN_API MdTester *Create()
{
    return new NonUniqueName();
}

File: plugins/Tester/nonUniqueName.py
import medic
from maya import OpenMaya


class NonUniqueName(medic.PyTester):
    def __init__(self):
        super(NonUniqueName, self).__init__()

    def Name(self):
        return "NonUniqueName"

    def Description(self):
        return "Non unique name(s) exists"

    def Match(self, node):
        return node.object().hasFn(OpenMaya.MFn.kDagNode)

    def IsFixable(self):
        return True

    def GetParameters(self):
        con = medic.ParamContainer.Create()
        p = medic.Parameter.Create("pattern", "Rename Pattern", medic.Types.String, "")
        con.append(p)
        return con

    def testNode(self, node):
        if node.dg().hasUniqueName():
            return None

        return medic.PyReport(node)

    def __setNewName(self, node, base):
        i = 1
        sel = OpenMaya.MSelectionList()
        while (True):
            try:
                sel.add("%s%d" % (base, i))
            except:
                break
            i += 1

        node.dg().setName("%s%d" % (base, i))

    def fix(self, report, params):
        node = report.node()
        if node.dg().isFromReferencedFile():
            return False

        rename_target = node

        if not node.dg().hasUniqueName():
            if node.object().hasFn(OpenMaya.MFn.kShape) and node.isDag():
                parents = node.parents()
                if parents:
                    parent = parents[0]
                    if not parent.dg().hasUniqueName():
                        rename_target = parent

            pattern = params.get("pattern")
            base = ""

            if pattern:
                base = pattern

            else:
                base = rename_target.name().split("|")[-1] + "_"

            self.__setNewName(rename_target, base)

        return True


def Create():
    return NonUniqueName()
